text/0000-api-evolution.md

```diff
@@ -0,0 +1,798 @@
+
+Making this work well in practice will likely require some infrastructure work
+along the lines of
+[this RFC issue](https://github.com/rust-lang/rfcs/issues/1029)
+
+## Mitigation for minor changes
+
+### The Crater tool
+
+@brson has been hard at work on a tool called "Crater" which can be used to
+exercise changes on the entire crates.io ecosystem, looking for
+regressions. This tool will be indispensable when weighing the costs of a minor
+change that might cause some breakage -- we can actually gauge what the breakage
+would look like in practice.
+
+While this would, of course, miss code not available publicly, the hope is that
+code on crates.io is a broadly representative sample, good enough to turn up
+problems.
+
+Any breaking, but minor change to the standard library must be evaluated through
+Crater before being committed.
+
+### Nightlies
+
+One line of defense against a "minor" change causing significant breakage is the
+nightly release channel: we can get feedback about breakage long before it makes
+even into a beta release. And of course the beta cycle itself provides another
+line of defense.
+
+### Elaborated source
+
+When compiling upstream dependencies, it is possible to generate an "elaborated"
+version of the source code where all dispatch is resolved to explicit UFCS form,
+all types are annotated, and all glob imports are replaced by explicit imports.
+
+This fully-elaborated form is almost entirely immune to breakage due to any of
+the "minor changes" listed above.
+
+You could imagine Cargo storing this elaborated form for dependencies upon
+compilation. That would in turn make it easy to update Rust, or some subset of
+dependencies, without breaking any upstream code (even in minor ways). You would
+be left only with very small, local changes to make to the code you own.
+
+While this RFC does not propose any such tooling change right now, the point is
+mainly that there are a lot of options if minor changes turn out to cause
+breakage more often than anticipated.
+
+### Trait item renaming
+
+One very useful mechanism would be the ability to import a trait while renaming
+some of its items, e.g. `use some_mod::SomeTrait with {foo_method as bar}`. In
+particular, when methods happen to conflict across traits defined in separate
+crates, a user of the two traits could rename one of the methods out of the way.
+
+## Thoughts on possible language changes (unofficial)
+
+The following is just a quick sketch of some focused language changes that would
+help our API evolution story.
+
+**Glob semantics**
+
+As already mentioned, the fact that glob imports currently allow *no* shadowing
+is deeply problematic: in a technical sense, it means that the addition of *any*
+public item can break downstream code arbitrarily.
+
+It would be much better for API evolution (and for ergonomics and intuition) if
+explicitly-defined items trump glob imports. But this is left to a future RFC.
```

Comment:
Is there any possibility that the changes to name resolution from such an RFC would be a breaking change? Are we allowing ourselves that possibility because "we intended it to be this way"?